### ISSUE TITLE
:D subset type constraints erroneously typecheck True against themselves

### ISSUE THREAD

**Issue by vrurg** [rakudo/rakudo#2166]
## The Problem

When an attribute or a scalar is declared with a type, and `:D`, and `where` `:D` is then disrepected making `Nil` assignment possible.

The following code:

```
my Int:D $foo where * >= 0 = 0; $foo = Nil; say $foo
```

clearly demonstrates the issue.

## Expected Behavior

The following output would be expected here:

```
Type check failed in assignment to $foo; expected type Int:D cannot be itself (perhaps Nil was assigned to a :D which had no default?)
  in block <unit> at -e line 1
```

## Actual Behavior

Rakudo ouputs this:

```
(<anon>)
```

## Environment
* Operating system: macOS 10.13
* Compiler version (`perl6 -v`): This is Rakudo version 2018.06 built on MoarVM version 2018.06

**Comment by zoffixznet** [comment]
Did a bit of digging on this and [this conditional](https://github.com/rakudo/rakudo/blob/e0a27ce3718fcb5bd0277b9760aea986c904c611/src/Perl6/Metamodel/BOOTSTRAP.nqp#L1587) is the one that OKs the default of the container.

So looks like it golfs down to the `nqp::istype` op OKing things that it probably shouldn't:

```
<Zoffix_> m: my Int:D $foo = 42; use nqp; dd nqp::istype($foo.VAR.default, $foo.VAR.of)
<camelia> rakudo-moar e0a27ce37: OUTPUT: «0␤»
<Zoffix_> m: my Int:D $foo where {True} = 42; use nqp; dd nqp::istype($foo.VAR.default, $foo.VAR.of)
<camelia> rakudo-moar e0a27ce37: OUTPUT: «1␤»
```

And it also affects `subsets`, which typecheck against its own typeobject:

```
<Zoffix_> m: subset Foo of Int:D where {True};  use nqp; dd nqp::istype(Foo, Foo)
<camelia> rakudo-moar e0a27ce37: OUTPUT: «1␤»
<Zoffix_> m: subset Foo of Int:D where {True};  use nqp; dd Foo ~~ Foo
<camelia> rakudo-moar e0a27ce37: OUTPUT: «Bool::True␤»
```

**Comment by vrurg** [comment]
Unless subset is having very special meaning, I don't think there is something wrong about `nqp::istype` matching a subset to itself. 

What is not clear to me is wether `Nil` being converted to the immediate type object of LHS is the right thing to do in this case. Perhaps it has to use the base type – `Int:D` in the above case?

**Comment by vrurg** [comment]
After some in-depth investigation it turns out that the problem has some conceptual aspect. The point is that `DefiniteHOW` and `SubsetHOW` are the two classes with `nominalizable` archetype. Both are not `nominal`. Where `DefiniteHOW` is handled correctly at the time of instantiating a new container descriptor and gets kinda nominalized (`maybe-definite-how-base` `World` method should've used `nominalize` method) `SubsetHOW` remains untouched resulting in it's been the default value for a scalar without `is default` trait applied.

I consider this incorrect because 

- this causes the erroneous behavior reported in this ticket
- only nominal type can be used as default `default`. A user could define whatever he wants to using `is default` but the language must not.

Hence, my conclusion is that all nominalizable types must be nominalized down to their respective base classes. 

Yet, this causes an unwanted side effect for subsets. Where for a definite type we know it has to be initialized, for a subset there is no such imperative. Therefore a declaration like:

```
my Int $a where { $_.defined and $_ > 42 };
```

is totally valid. But it would fail if we implement nominalization right away because `Int` wouldn't pass the `where` constraint!

I'm currently trying to figure out a way around this. So far, the only solution which comes to my mind is to introduce an additional flag attribute on `ContainerDescriptor` which would signalize if the `default` is implicit or explicit. The former would be the default state after the descriptor construction. The latter would be the case after calling `set_default` method. This way I hope to preserve the backward compatibility at enough level as to pass the spectest as is.

Any considerations are very much welcome.

**Comment by vrurg** [comment]
Closed via rakudo/rakudo#2946
